### PR TITLE
BF+RF: remove code duplication of (ensure_|anything2)bool and make more consistent with git

### PR DESCRIPTION
**Pull request by yarikoptic**
This is "inspired" by [a recent annex issue discussion](https://git-annex.branchable.com/bugs/does_not_understand___34__yes__34___as_boolean_true_value_for_autoenable/?updated#comment-df028d791f7220741cf9837462618043), where I detected that we were not inline with git-config handling of strings containing negative integers as `True`.  Then I realized that we have two implementations intended for the same thing, which we use pretty much similarly in different points. `ensure_bool` is used in
```python
$> git grep sure_bool | grep -v test_
datalad/core/distributed/clone.py:    assure_bool,
datalad/core/distributed/clone.py:            sr_autoenable = assure_bool(sr_autoenable)
```
and by `datalad-crawler` for handling crawler configuration settings.  `anything2bool` - by datalad.config.  So - both are used for the same purpose but (still) behave slightly different

<details>
<summary>comparison of ensure_bool (anything2bool was ok besides negative int) with git-config before this PR</summary> 

```shell
$> for v in true yes 1 y 200 '' ' ' false no n 0 -1 0.00 0.1 string 1.0; do echo -n "$v="; git -c "sec.blah=$v" config --bool sec.blah; python -c "from datalad.utils import ensure_bool; print(ensure_bool('$v'))" ; echo; done
true=true
True

yes=true
True

1=true
True

y=fatal: bad numeric config value 'y' for 'sec.blah': invalid unit
True

200=true
True

=false
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 781, in ensure_bool
    raise ValueError("Do not know how to treat %r as a boolean" % s)
ValueError: Do not know how to treat '' as a boolean

 =fatal: bad numeric config value ' ' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 781, in ensure_bool
    raise ValueError("Do not know how to treat %r as a boolean" % s)
ValueError: Do not know how to treat ' ' as a boolean

false=false
False

no=false
False

n=fatal: bad numeric config value 'n' for 'sec.blah': invalid unit
False

0=false
False

-1=true
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 781, in ensure_bool
    raise ValueError("Do not know how to treat %r as a boolean" % s)
ValueError: Do not know how to treat '-1' as a boolean

0.00=fatal: bad numeric config value '0.00' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 781, in ensure_bool
    raise ValueError("Do not know how to treat %r as a boolean" % s)
ValueError: Do not know how to treat '0.00' as a boolean

0.1=fatal: bad numeric config value '0.1' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 781, in ensure_bool
    raise ValueError("Do not know how to treat %r as a boolean" % s)
ValueError: Do not know how to treat '0.1' as a boolean

string=fatal: bad numeric config value 'string' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 781, in ensure_bool
    raise ValueError("Do not know how to treat %r as a boolean" % s)
ValueError: Do not know how to treat 'string' as a boolean

1.0=fatal: bad numeric config value '1.0' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 781, in ensure_bool
    raise ValueError("Do not know how to treat %r as a boolean" % s)
ValueError: Do not know how to treat '1.0' as a boolean
```
</details>

<details>
<summary>comparison of anythin2bool with git-config on typical values after this PR</summary> 

```shell
$> for v in true yes 1 y 200 '' ' ' false no n 0 -1 0.00 0.1 string 1.0; do echo -n "$v="; git -c "sec.blah=$v" config --bool sec.blah; python -c "from datalad.config import anything2bool; print(anything2bool('$v'))" ; echo; done           
true=true
True

yes=true
True

1=true
True

y=fatal: bad numeric config value 'y' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/config.py", line 125, in anything2bool
    return ensure_bool(val, short=False)
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 792, in ensure_bool
    raise ValueError("Do not know how to treat %s as a boolean" % repr(s))
ValueError: Do not know how to treat 'y' as a boolean

200=true
True

=false
False

 =fatal: bad numeric config value ' ' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/config.py", line 125, in anything2bool
    return ensure_bool(val, short=False)
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 792, in ensure_bool
    raise ValueError("Do not know how to treat %s as a boolean" % repr(s))
ValueError: Do not know how to treat ' ' as a boolean

false=false
False

no=false
False

n=fatal: bad numeric config value 'n' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/config.py", line 125, in anything2bool
    return ensure_bool(val, short=False)
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 792, in ensure_bool
    raise ValueError("Do not know how to treat %s as a boolean" % repr(s))
ValueError: Do not know how to treat 'n' as a boolean

0=false
False

-1=true
True

0.00=fatal: bad numeric config value '0.00' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/config.py", line 125, in anything2bool
    return ensure_bool(val, short=False)
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 792, in ensure_bool
    raise ValueError("Do not know how to treat %s as a boolean" % repr(s))
ValueError: Do not know how to treat '0.00' as a boolean

0.1=fatal: bad numeric config value '0.1' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/config.py", line 125, in anything2bool
    return ensure_bool(val, short=False)
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 792, in ensure_bool
    raise ValueError("Do not know how to treat %s as a boolean" % repr(s))
ValueError: Do not know how to treat '0.1' as a boolean

string=fatal: bad numeric config value 'string' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/config.py", line 125, in anything2bool
    return ensure_bool(val, short=False)
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 792, in ensure_bool
    raise ValueError("Do not know how to treat %s as a boolean" % repr(s))
ValueError: Do not know how to treat 'string' as a boolean

1.0=fatal: bad numeric config value '1.0' for 'sec.blah': invalid unit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/yoh/proj/datalad/datalad-maint/datalad/config.py", line 125, in anything2bool
    return ensure_bool(val, short=False)
  File "/home/yoh/proj/datalad/datalad-maint/datalad/utils.py", line 792, in ensure_bool
    raise ValueError("Do not know how to treat %s as a boolean" % repr(s))
ValueError: Do not know how to treat '1.0' as a boolean

```
</details>

- [ ] decide if I should just make `short=False` or remove altogether, thus requiring users to say full "yes".  If I run into any stored configuration in the crawler, it should be an easy fix.  So I am leaning towards that

FWIW, this PR is nohow helps #2628 but would change exception type from TypeError to ValueError.

### CODE REVIEW

File: datalad/tests/test_config.py [hunk 1]
```diff
@@ -102,7 +102,7 @@ def test_something(path, new_home):
     assert_equal(cfg.getint('something', 'myint'), 3)
     assert_equal(cfg.getbool('something', 'myint'), True)
     assert_equal(cfg.getbool('doesnot', 'exist', default=True), True)
-    assert_raises(TypeError, cfg.getbool, 'something', 'user')
+    assert_raises(ValueError, cfg.getbool, 'something', 'user')
```
Comment (kyleam) [inline datalad/tests/test_config.py:105]: A change in the exception raised should not be going into maint.

Comment (yarikoptic) [reply datalad/tests/test_config.py:105]: Words of wisdom ;) agreed -- I will adjust target to be `master`

**Comment by yarikoptic** [comment]
docs build failure is "interesting":
```
reading sources... [100%] related

/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/site-packages/sphinx/application.py:993: RemovedInSphinx40Warning: app.add_lexer() API changed; Please give lexer class instead instance
  RemovedInSphinx40Warning)

Warning, treated as error:
/home/runner/work/datalad/datalad/docs/source/generated/cfginfo/dataset.rst:2:duplicate term description of datalad.locations.dataset-procedures, other instance in config
make: *** [html] Error 2
Makefile:55: recipe for target 'html' failed
```
and seems came about a few days ago and currently present in other PRs: https://github.com/datalad/datalad/actions?query=workflow%3ADocs

**Comment by yarikoptic** [comment]
oh, we didn't have `maint` merged into master recently, so I will just rebase.

While at it - I will also just kill that `short` flag, removing handling of `y` and `n`. IMHO better to stay consistent with git.

**Comment by yarikoptic** [comment]
Ok, then I will add an option cast=True to ensure_bool with anything2bool setting it to False

**Comment by yarikoptic** [comment]
> > I think it'd be reasonable for a caller to expect a general utility function called ensure_bool to be able to handle an object according to its __bool__
> Ok, then I will add an option cast=True to ensure_bool with anything2bool setting it to False

hm... thought about that.  The fact is that `ensure_bool('some magical string values') == False` already undermines the magical power expectation of the "general utility".  So not really sure if retaining casting is making life easier.  But pushed e3efa7bb216553c8a2266d606a4709649eedc5d5

**Comment by kyleam** [comment]
> hm... thought about that. The fact is that `ensure_bool('some magical string values') == False` already undermines the magical power expectation of the "general utility".

Yeah, fair point.  It maybe comes down to `ensure_bool` being a bad name for this, especially when taken alongside our other `ensure_*` variants.  Still, I find your example less surprising than having an object with a custom `__bool__` not being respected.

> So not really sure if retaining casting is making life easier. But pushed [e3efa7b](https://github.com/datalad/datalad/commit/e3efa7bb216553c8a2266d606a4709649eedc5d5)

As I said originally, I don't feel strongly about it, though I lean towards avoiding the stricter behavior given there's a chance of breaking third-party code.  But if you disagree, there's no need to go with something you don't want here.

Side note: As I was looking through the code and thinking about this again, I noticed that EnsureBool has mostly duplicated logic, with a few twists.

**Comment by mih** [comment]
Last push >6 month ago. Closing do to inactivity.

**Pull request closed without merging**